bokehjs/src/lib/models/ranges/factor_range.ts

```diff
@@ -186,119 +239,99 @@ export class FactorRange extends Range {
-        const y0 = mapping.get(f0)
-        return y0 != null ? y0.value : NaN
+        return map_l1(x, this.mapping as L1Mapping)
       }
       case 2: {
-        const [f0, f1] = x
-        const mapping = this._mapping as L2Mapping
-        const y0 = mapping.get(f0)
-        if (y0 != null) {
-          const y1 = y0.mapping.get(f1)
-          if (y1 != null) {
-            return y1.value
-          }
-        }
-        return NaN
+        return map_l2(x, this.mapping as L2Mapping)
       }
       case 3: {
-        const [f0, f1, f2] = x
-        const mapping = this._mapping as L3Mapping
-        const y0 = mapping.get(f0)
-        if (y0 != null) {
-          const y1 = y0.mapping.get(f1)
-          if (y1 != null) {
-            const y2 = y1.mapping.get(f2)
-            if (y2 != null) {
-              return y2.value
-            }
-          }
-        }
-        return NaN
+        return map_l3(x, this.mapping as L3Mapping)
       }
     }
   }
 
-  // convert a string factor into a synthetic coordinate
+  protected _lookup_value(x: BoxedFactor): number {
+    return this._lookup_entry(x)?.value ?? NaN
+  }
+
+  // convert a categorical factor into a synthetic coordinate
   synthetic(x: FactorLike): number {
     if (isNumber(x)) {
       return x
     }
 
     if (isString(x)) {
-      return this._lookup([x])
+      return this._lookup_value([x])
     }
 
-    let offset = 0
-    const off = x[x.length-1]
-    if (isNumber(off)) {
-      offset = off
-      x = x.slice(0, -1) as BoxedFactor
+    const offset = x.at(-1)
+    if (isNumber(offset)) {
+      return this._lookup_value(x.slice(0, -1) as BoxedFactor) + offset
     }
 
-    return this._lookup(x as BoxedFactor) + offset
+    return this._lookup_value(x as BoxedFactor)
   }
 
-  // convert an array of string factors into synthetic coordinates
+  // convert an array of categorical factors into synthetic coordinates
   v_synthetic(xs: Arrayable<number | Factor | [string] | OffsetFactor>): ScreenArray {
-    const n = xs.length
-    const array = new ScreenArray(n)
-    for (let i = 0; i < n; i++) {
-      array[i] = this.synthetic(xs[i])
+    return ScreenArray.from(xs, (x) => this.synthetic(x))
+  }
+
+  // convert a synthetic coordinate into a categorical factor
+  factor(x: number): Factor | null {
+    for (const f of this.factors) {
+      const v = this.synthetic(f)
+      if (x >= (v-0.5) && x < (v+0.5)) {
+        return f
+      }
     }
-    return array
+    return null
+  }
+
+  protected _compute_bounds(inner_padding: number): [number, number] {
+    const interval = this.factors.length + inner_padding
+    const padding = (() => {
+      switch (this.range_padding_units) {
+        case "percent": {
+          return interval * this.range_padding / 2
+        }
+        case "absolute": {
+          return this.range_padding
+        }
+      }
+    })()
+    // XXX "0 - padding" is a workaround for an assertions bug
```

Comment:
This is the tester bug I mentioned in the other issue

Reply:
We rely on `Object.is()` which is great for NaNs, but apparently treats 0 and -0 differently than `===` does.